### PR TITLE
Depth map management

### PR DESCRIPTION
**Pull request by guycalledfrank**
Moved all depth map management to camera/renderer instead of being scattered in post effects and particles, leading to possible conflicts.
- scene_camera has methods requestDepthMap and releaseDepthMap which increment/decrement a reference counter. The depth map will be created if ref count > 0 and destroyed otherwise.
- particle emitters with depth softening request depth map
- post effects request depth map if needsDepthBuffer is set (tested DoF effect)
- renderer now provides global uDepthMap (if depth rendering enabled) and uScreenSize (always) uniforms.

Additional:
- device now silently ignores unset samplers, because it's better than crash.
- fixed renderer.setCamera not selecting color channels to clear (reason of background color flickering with fixed particles)

### CODE REVIEW

File: src/posteffect/posteffect_posteffectqueue.js [hunk 1]
```diff
@@ -226,10 +206,11 @@ pc.extend(pc, function () {
                         var len = effects.length;
                         if (len) {
                             camera.renderTarget = effects[0].inputTarget;
-                            this._setDepthTarget(this.depthTarget);
+                            this.depthTarget = this.camera.camera._depthTarget;
```
Comment (daredevildave) [inline src/posteffect/posteffect_posteffectqueue.js]: Accessing private members?

File: src/scene/scene_forwardrenderer.js [hunk 2]
```diff
@@ -631,7 +636,31 @@ pc.extend(pc, function () {
             drawCalls.sort(sortDrawCalls);
 
             // Render a depth target if the camera has one assigned
-            if (camera._depthTarget) {
+            if (camera._renderDepthRequests) {
```
Comment (daredevildave) [inline src/scene/scene_forwardrenderer.js]: Again, leading _ should indicate that a member is for private internal use.

We shouldn't be accessing them outside of the class.

File: src/posteffect/posteffect_posteffectqueue.js (hunk 1 above)
Comment (guycalledfrank) [inline src/posteffect/posteffect_posteffectqueue.js]: Who cares? Look at the original source: this posteffects file CREATED the _depthTarget on camera before, and ForwardRenderer was reading it. So I didn't make it worse.

File: src/scene/scene_forwardrenderer.js (hunk 2 above)
Comment (guycalledfrank) [inline src/scene/scene_forwardrenderer.js]: Again it's not worse than the original code
https://github.com/playcanvas/engine/blob/master/src/scene/scene_forwardrenderer.js#L634

Comment (vkalpias) [inline src/scene/scene_forwardrenderer.js]: Well we should at least make it better right?

Comment (daredevildave) [inline src/scene/scene_forwardrenderer.js]: Fair point, but we should be trying to make the code better :-)

In this case, shouldn't we just have a method which is `camera.hasDepthMap` or whatever it should be called?

Comment (guycalledfrank) [inline src/scene/scene_forwardrenderer.js]: Renderer also reads the depth map.

File: src/posteffect/posteffect_posteffectqueue.js (hunk 1 above)
Comment (daredevildave) [inline src/posteffect/posteffect_posteffectqueue.js]: Should we be exposing the depthTarget property properly?

Comment (guycalledfrank) [inline src/posteffect/posteffect_posteffectqueue.js]: Maybe, at least as read-only.

File: src/scene/scene_forwardrenderer.js (hunk 2 above)
Comment (daredevildave) [inline src/scene/scene_forwardrenderer.js]: The point of this is communication of API though. If I see a method that is documented, then I know I'm safe to use it. If I'm searching for properties with a leading underscore then I know I shouldn't use it.

Which is this? Should users of the camera node have access to this property or is it secret only for internal use.

It looks like authors of post effects will need to use it.

Comment (guycalledfrank) [inline src/scene/scene_forwardrenderer.js]: Original post effect system internally sets depthMap variable for each effect that requires depth though.

**Pull request merged**